### PR TITLE
Logarithmic merge strategy

### PR DESCRIPTION
**Pull request by jotare**
### Description
Describe the proposed changes made in this PR.

### How was this PR tested?
Describe how you tested this PR.

### CODE REVIEW

File: nidx/src/maintenance/scheduler.rs [hunk 1]
```diff
@@ -167,37 +170,150 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
...
+    min_number_of_segments: usize,
+
+    /// Max number of records for a segment to be in the top bucket, i.e.,
+    /// elegible for merge. Once a segment becomes bigger, it won't be merged
+    /// anymore
+    top_bucket_max_records: usize,
+
+    /// Max number of records for a segment to be considered in the bottom
+    /// bucket. Segments with fewer records won't be further splitted in buckets
+    bottom_bucket_threshold: u32,
```
Comment (javitonino) [inline nidx/src/maintenance/scheduler.rs:227]: It's weird to have multiple integer formats here, maybe standardize everything to usize?

File: nidx/src/maintenance/scheduler.rs [hunk 2]
```diff
@@ -167,37 +170,150 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
...
+                    current_max_size_log = segment_size_log;
+                }
+
+                current_bucket.push(segment_id);
+            }
+            buckets.push(current_bucket);
+
+            for segments in buckets {
+                if segments.len() >= self.min_number_of_segments {
+                    println!("Merge job for bucket: {segments:?}");
```
Comment (javitonino) [inline nidx/src/maintenance/scheduler.rs:309]: Maybe switch to an info!/debug! line instead, e.g:

```suggestion
                    debug!(?segments, "Creating merge job for bucket");
```

File: nidx/src/maintenance/scheduler.rs [hunk 3]
```diff
@@ -167,37 +170,150 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
...
+                    current_bucket = vec![];
+                    current_max_size_log = segment_size_log;
+                }
+
+                current_bucket.push(segment_id);
+            }
+            buckets.push(current_bucket);
+
+            for segments in buckets {
+                if segments.len() >= self.min_number_of_segments {
```
Comment (javitonino) [inline nidx/src/maintenance/scheduler.rs:308]: We are missing merging for deletions. It would be another condition here checking to allow merging a bucket with a single segment, if the numbers of deleted keys on seqs between  >segment_seq and <=last_indexed_seq exceeds a threshold.

Can be a separate PR if you prefer, but let's not forget it.

File: nidx/src/maintenance/scheduler.rs [hunk 4]
```diff
@@ -214,6 +330,148 @@ mod tests {
...
         use crate::metadata::{Index, IndexId, IndexKind, NidxMetadata, Shard};
 
+        fn merge_scheduler() -> MergeScheduler {
+            MergeScheduler::from_settings(&MergeSettings {
+                min_number_of_segments: 3,
+                ..Default::default()
+            })
+        }
+
+        async fn create_segments(pool: &sqlx::PgPool, index: &Index, segment_records: &[i64]) -> anyhow::Result<Seq> {
```
Comment (javitonino) [inline nidx/src/maintenance/scheduler.rs:340]: nit: I would like if this function name explained why it's receiving a vec! of records, something like `create_segments_with_num_records`

**Review by javitonino — APPROVED** [review]

File: nidx/src/maintenance/scheduler.rs (hunk 2 above)
Comment (jotare) [reply nidx/src/maintenance/scheduler.rs:309]: It was a leftover while debugging. I'll change it for a debug

File: nidx/src/maintenance/scheduler.rs (hunk 3 above)
Comment (jotare) [reply nidx/src/maintenance/scheduler.rs:314]: I'll implement it in another PR
